### PR TITLE
Remove prefix for body model binding(#1694)

### PR DESCRIPTION
**Pull request by kirthik**
Fix for #1694

### CODE REVIEW

**Comment by kirthik** [comment]
@harshgMSFT - Could you please take a look?

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    /// <summary>
+    /// Utility class for clearing entries in <see cref="ModelStateDictionary"/>.
+    /// </summary>
+    public static class ModelStateDictionaryUtility
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: This kind of functionality should just be instance methods on MSD.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            [NotNull] string name)
+        {
+            if (modelState.ContainsKey(name))
+            {
+                modelState[name].Errors.Clear();
+                modelState[name].ValidationState = ModelValidationState.Unvalidated;
+            }
+        }
+
+        private static bool IsSimpleType(Type type)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: Why is this there?

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 3]
```diff
@@ -1209,6 +1209,8 @@ public virtual bool TryValidateModel([NotNull] object model, string prefix)
                 throw new InvalidOperationException(message);
             }
 
+            ModelStateDictionaryUtility.ClearModelStateDictionary(ModelState, model.GetType());
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: wouldnt you need a similar call for ```APIController``` ?

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        /// <summary>
+        /// Clears <see cref="ModelStateDictionary"/>.
+        /// </summary>
+        /// <param name="modelState">The <see cref="ModelStateDictionary"/>.</param>
+        /// <param name="type">The <see cref="Type"/> of object whose properties entries have to be cleared from
+        /// <see cref="ModelStateDictionary"/>.</param>
+        public static void ClearModelStateDictionary([NotNull] ModelStateDictionary modelState, [NotNull] Type type)
+        {
+            var properties = type.GetProperties();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: We really really really want to avoid just dropping down to reflection, even when it helps us solve a problem - because it creates another one... now we have reflection code that we have to implement caching for.

What's probable the right change here is to do a prefix check for the model state key that we're clearing. That will also solve the problem for models that have indexing (array, dictionary, etc).

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 5]
```diff
@@ -60,7 +60,7 @@ public class BodyModelBinder : BindingSourceModelBinder
             }
 
             var model = await formatter.ReadAsync(formatterContext);
-            return new ModelBindingResult(model, bindingContext.ModelName, isModelSet: true);
+            return new ModelBindingResult(model, string.Empty, isModelSet: true);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: named parameters.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: Any time you find yourself reaching for a word like `Utility` or `Helper` it's usually a clue that there's a better option.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs (hunk 3 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: Also I thought the decision was to invalidate the tree for the model type _ONLY_ and not the entire state. The name of the method tells me that it will clear everything.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            if (properties.Count() > 0)
+            {
+                foreach (var p in properties)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: ```var property in```

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                {
+                    if (IsSimpleType(p.PropertyType))
+                    {
+                        ClearModelStateDictionaryEntry(modelState, p.Name);
+                    }
+                    else
+                    {
+                        ClearModelStateDictionary(modelState, p.PropertyType);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: Isnt it possible to get an infinite loop? consider a type 
```class Person
{
       public Person Parent {get; set;}
}
```

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                    }
+                }
+            }
+            return;
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: not needed.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: :+1 lets not add helper classes.

**Comment by rynowak** [comment]
I don't understand why these changes go together. #2071 is not an approved bug

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 2 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: Please use the common ```TypeHelper```
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Common/TypeHelper.cs

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 9]
```diff
@@ -38,17 +38,17 @@ public class TryValidateModelTest
...
-            Assert.Equal("The Category field is required.", json["product.Category"]);
-            Assert.Equal("The field Contact Us must be a string with a maximum length of 20."+
-                "The field Contact Us must match the regular expression '^[0-9]*$'.", json["product.Contact"]);
+            Assert.Equal(4, json.Count);
             Assert.Equal("CompanyName cannot be null or empty.", json["CompanyName"]);
             Assert.Equal("The field Price must be between 20 and 100.", json["Price"]);
             Assert.Equal("The Category field is required.", json["Category"]);
             Assert.Equal("The field Contact Us must be a string with a maximum length of 20."+
                 "The field Contact Us must match the regular expression '^[0-9]*$'.", json["Contact"]);
+            //Assert.Equal("CompanyName cannot be null or empty.", json["CompanyName"]);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: why is this commented? should you just remove ?

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 10]
```diff
@@ -7,6 +7,7 @@
 using System.Linq;
 using Microsoft.AspNet.Mvc;
 using ValidationWebSite.ViewModels;
+using Microsoft.AspNet.Mvc.ModelBinding;
```
Comment (harshgMSFT) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: why is this needed?

**Comment by harshgMSFT** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 1 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: Ok. I will move the code to MSD class.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 2 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: I din't know this existed. I will use this instead.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs (hunk 3 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: Yes I am just clearing the entries for properties in model. Will use a better name for the method.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs (hunk 4 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: @rynowak  - Where do I get the prefix? Do you mean I should do something similar to what we are doing in `DefaultObjectValidator's` `ValidateProperties` method to get model state keys/prefix?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: It would seem that it would have to be passed in right? You can pass in a prefix to TryUpdateModel/TryValidateModel. If one isn't provided then it's the empty string

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: I realize the problem that you'll run into with this is when the prefix is empty you'd end up clearing too much. Maybe we need to use ModelMetadata here.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: MSD interactions normally start with the full name of the entry.  why not pass in `string fullName` or (to match other MSD methods) `string key` instead of `Type type`?  clear everything from that `key` down.  worst case, iterate through the entire dictionary and remove information from all entries that have keys matching `"key|(key\[.*)|(key\..*)"` but there's likely more efficient ways to do it.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionaryUtility.cs]: :100: 

Thanks @dougbu that's what I was trying to say

**Comment by kirthik** [comment]
@harshgMSFT : Updated PR. Could you please take a look?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 11]
```diff
@@ -69,10 +69,10 @@ public class BodyModelBinder : BindingSourceModelBinder
             {
                 model = GetDefaultValueForType(bindingContext.ModelType);
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Not specific to this PR, but is this value actually used?
Same remark for the `_bodyValidationExcludeFiltersProvider`, which is imported but not used.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 12]
```diff
@@ -57,7 +57,7 @@ public class BodyModelBinder : BindingSourceModelBinder
                 var unsupportedContentType = Resources.FormatUnsupportedContentType(
                     bindingContext.OperationBindingContext.HttpContext.Request.ContentType);
                 bindingContext.ModelState.AddModelError(bindingContext.ModelName, unsupportedContentType);
-                return new ModelBindingResult(null, bindingContext.ModelName, isModelSet: false);
+                return new ModelBindingResult(model: null, key: string.Empty, isModelSet: false);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Add a comment why we are not using the bindingContext.ModelName.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 13]
```diff
@@ -57,7 +57,7 @@ public class BodyModelBinder : BindingSourceModelBinder
                 var unsupportedContentType = Resources.FormatUnsupportedContentType(
                     bindingContext.OperationBindingContext.HttpContext.Request.ContentType);
                 bindingContext.ModelState.AddModelError(bindingContext.ModelName, unsupportedContentType);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Add tests for this.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs (hunk 12 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Lets return bindingContext.ModelName when we fail, In this case since we ourselves are adding model state error we should return that key.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 14]
```diff
@@ -69,10 +69,10 @@ public class BodyModelBinder : BindingSourceModelBinder
             {
                 model = GetDefaultValueForType(bindingContext.ModelType);
                 bindingContext.ModelState.AddModelError(bindingContext.ModelName, ex);
-                return new ModelBindingResult(model: null, key: bindingContext.ModelName, isModelSet: false);
+                return new ModelBindingResult(model: null, key: string.Empty, isModelSet: false);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Lets return bindingContext.ModelName when we fail, In this case since we ourselves are adding model state error we should return that key.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 15]
```diff
@@ -69,10 +69,10 @@ public class BodyModelBinder : BindingSourceModelBinder
...
             }
 
-            return new ModelBindingResult(model, bindingContext.ModelName, isModelSet: true);
+            return new ModelBindingResult(model, key: string.Empty, isModelSet: true);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Add a comment why this is not bindingContext.ModelName.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 16]
```diff
@@ -37,12 +37,7 @@ public class TryValidateModelTest
             // Assert
             var body = await response.Content.ReadAsStringAsync();
             var json = JsonConvert.DeserializeObject<Dictionary<string, string>>(body);
-            Assert.Equal(8, json.Count);
-            Assert.Equal("CompanyName cannot be null or empty.", json["product.CompanyName"]);
-            Assert.Equal("The field Price must be between 20 and 100.", json["product.Price"]);
-            Assert.Equal("The Category field is required.", json["product.Category"]);
-            Assert.Equal("The field Contact Us must be a string with a maximum length of 20." +
-                "The field Contact Us must match the regular expression '^[0-9]*$'.", json["product.Contact"]);
+            Assert.Equal(4, json.Count);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: Why are there only 4 entries as opposed to 8?

**Comment by harshgMSFT** [comment]
:watch:  on minor changes and few tests.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs (hunk 16 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: before this change, validation after model binding added 4 errors with `product` prefix and then a call to `TryValidate` added the 4 errors without prefix. After this change, we no longer get the prefix errors and `TryValidate' clears the errors in ModelState and re-writes them

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs (hunk 12 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Ok.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs (hunk 13 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Will do

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs (hunk 14 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Ok

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs (hunk 15 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Will do

**Comment by kirthik** [comment]
@harshgMSFT - Updated PR. Please take a look.

**Comment by harshgMSFT** [comment]
:watch on one more test.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs (hunk 16 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: :+1:

**Comment by harshgMSFT** [comment]
:shipit:

**Comment by kirthik** [comment]
Fix checked in - 2b246e7acc50a4a3fc19d5e36578e194d1bd94e3
